packages/react-instantsearch-core/src/widgets/__tests__/DynamicWidgets.test.tsx

```diff
@@ -0,0 +1,331 @@
+              },
+            },
+          ],
+        },
+      ];
+
+      const Component = ({
+        result,
+      }: {
+        result: React.ComponentProps<typeof InstantSearch>['resultsState'];
+      }) => (
+        <InstantSearch
+          searchClient={searchClient}
+          indexName="test"
+          resultsState={result}
+        >
+          <DynamicWidgets>
+            <RefinementList attribute="test1" />
+            <HierarchicalMenu attributes={['test2', 'test3']} />
+          </DynamicWidgets>
+        </InstantSearch>
+      );
+      {
+        const { container } = render(
+          <Component
+            // @ts-ignore resultsState in InstantSearch is typed wrongly to deal with multi-index
+            result={RESULTS[0]}
+          />
+        );
+
+        expect(container).toMatchInlineSnapshot(`
+          <div>
+            RefinementList(test1)
+            HierarchicalMenu(test2,test3)
+          </div>
+        `);
+      }
+      {
+        const { container } = render(
+          <Component
+            // @ts-ignore resultsState in InstantSearch is typed wrongly to deal with multi-index
+            result={RESULTS[1]}
+          />
+        );
+
+        expect(container).toMatchInlineSnapshot(`
+          <div>
+            HierarchicalMenu(test2,test3)
+            RefinementList(test1)
+          </div>
+        `);
+      }
+      {
+        const { container } = render(
+          <Component
+            // @ts-ignore resultsState in InstantSearch is typed wrongly to deal with multi-index
+            result={RESULTS[2]}
+          />
+        );
+
+        expect(container).toMatchInlineSnapshot(`
+          <div>
+            RefinementList(test1)
+          </div>
+        `);
+      }
+    });
+
+    test('renders items in panel', () => {
+      const searchClient = createSearchClient();
+
+      const { container } = render(
+        <InstantSearch
+          searchClient={searchClient}
+          indexName="test"
+          // @ts-ignore resultsState in InstantSearch is typed wrongly to deal with multi-index
+          resultsState={resultsState}
+        >
+          <DynamicWidgets transformItems={() => ['test1', 'test3']}>
+            <RefinementList attribute="test1" />
+            <RefinementList attribute="test2" />
+            <Panel>
+              <RefinementList attribute="test3" />
+            </Panel>
+            <Panel>
+              <RefinementList attribute="test4" />
+            </Panel>
+          </DynamicWidgets>
+        </InstantSearch>
+      );
+
+      expect(container).toMatchInlineSnapshot(`
+        <div>
+          RefinementList(test1)
+          <div
+            class="ais-Panel"
+          >
+            <div
+              class="ais-Panel-body"
+            >
+              RefinementList(test3)
+            </div>
+          </div>
+        </div>
+      `);
+    });
+
+    test("does not render items that aren't directly in children", () => {
```

Comment:
Can we add a limitation section in the PR description that mentions that it doesn't support "proxy components"?

```jsx
function ProxyComponent({ attribute }) {
  return <RefinementList attribute={attribute} />;
}

function App() {
  return (
    <DynamicWidgets>
      <ProxyComponent attribute="brand" />
    </DynamicWidgets>
  );
}
```

It'll be easier to track later.